org.osgi.test.junit5.cm/src/main/java/org/osgi/test/junit5/cm/AnnotationUtil.java

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+ * Copyright (c) Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *******************************************************************************/
+
+package org.osgi.test.junit5.cm;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import org.osgi.test.common.annotation.config.WithConfiguration;
+import org.osgi.test.common.annotation.config.WithConfigurations;
+import org.osgi.test.common.annotation.config.WithFactoryConfiguration;
+import org.osgi.test.common.annotation.config.WithFactoryConfigurations;
+
+class AnnotationUtil {
+
+	private AnnotationUtil() {}
+
+	private static Predicate<Annotation> ANY_WITH_CONF_ANNNOTATIONS = t -> t.annotationType()
+		.equals(WithFactoryConfiguration.class)
+		|| t.annotationType()
+			.equals(WithFactoryConfigurations.class)
+		|| t.annotationType()
+			.equals(WithConfiguration.class)
+		|| t.annotationType()
+			.equals(WithConfigurations.class);
+
+	public static List<Annotation> findAllConfigAnnotations(AnnotatedElement annotatedElement) {
+		return findAllAnnotationsMatching(annotatedElement, ANY_WITH_CONF_ANNNOTATIONS);
+	}
+
+	private static List<Annotation> findAllAnnotationsMatching(AnnotatedElement annotatedElement,
+		Predicate<Annotation> test) {
+		final ArrayList<Annotation> found = new ArrayList<>();
+		findAllAnnotationsMatching(annotatedElement, test, found, new HashSet<>());
+		return found;
+	}
+
+	private static void findAllAnnotationsMatching(AnnotatedElement annotatedElement, Predicate<Annotation> test,
+		ArrayList<? super Annotation> found, Set<Entry<AnnotatedElement, Annotation>> visited) {
+		Annotation[] declaredAnnotations = annotatedElement.getDeclaredAnnotations();
+
+		for (Annotation ann : declaredAnnotations) {
+			if (isJavaLangAnnotation(ann) || isJunitAnnotation(ann)) {
+				continue;
+			}
+
+			if (visited.add(new AbstractMap.SimpleEntry<>(annotatedElement, ann))) {
```

Comment:
Shouldn't this just be a `Set<Annotation>` - do we really need to recursively process annotations that we've seen before (with the same values) if they're in a different place?

Reply:
yes. because we need them in the order.

if we have 
@Setup1
@Setup2
@Setup1


and @Setup2 overrides some parts of @Setup1

then its important to have the full scan

Reply:
that is testes by

```
@Test
	@ConfigSetup_B
	void testOrderUpperLatest_B() throws Exception {

		Configuration c = ConfigUtil.getConfigsByServicePid(ca, FACTORY_CONFIGURATION_PID + "~" + NAME);
		assertThat(c).isNotNull();
		DictionaryAssert.assertThat(c.getProperties())
			.containsEntry(KEY, ConfigSetup_2.VALUE);

		Configuration cs2 = ConfigUtil.getConfigsByServicePid(ca, CONFIGURATION_PID);
		assertThat(cs2).isNotNull();
		DictionaryAssert.assertThat(cs2.getProperties())
			.containsEntry(KEY, ConfigSetup_2.VALUE);

	}

	@Test
	@ConfigSetup_B
	@ConfigSetup_1
	void testOrderUpperLatest() throws Exception {

		Configuration c = ConfigUtil.getConfigsByServicePid(ca, FACTORY_CONFIGURATION_PID + "~" + NAME);
		assertThat(c).isNotNull();
		DictionaryAssert.assertThat(c.getProperties())
			.containsEntry(KEY, ConfigSetup_1.VALUE);

		Configuration cs2 = ConfigUtil.getConfigsByServicePid(ca, CONFIGURATION_PID);
		assertThat(cs2).isNotNull();
		DictionaryAssert.assertThat(cs2.getProperties())
			.containsEntry(KEY, ConfigSetup_1.VALUE);

	}


```